### PR TITLE
[runtime] Use MonoError for mono_object_new_pinned

### PR DESCRIPTION
**Pull request by ludovic-henry**

### CODE REVIEW

**Comment by ludovic-henry** [comment]
Depends on https://github.com/mono/mono/pull/2462

File: mono/metadata/object.c [hunk 1]
```diff
@@ -4510,12 +4510,14 @@ mono_object_new (MonoDomain *domain, MonoClass *klass)
...
 {
 	MONO_REQ_GC_UNSAFE_MODE;
 
 	MonoVTable *vtable;
 
+	mono_error_init (error);
+
 	vtable = mono_class_vtable (domain, klass);
 	if (!vtable)
 		return NULL;
```
Comment (kumpera) [inline mono/metadata/object.c]: This is wrong, we can't return NULL and not set MonoError.

Comment (ludovic-henry) [inline mono/metadata/object.c]: This is the same behavior as previously. The best fix for that case is to have and use `mono_class_vtable_checked`, but that's for another PR (https://github.com/mono/mono/pull/2455)

Comment (kumpera) [inline mono/metadata/object.c]: The existing behavior is wrong and we should not keep it that way. As we increase
our reliance on MonoError, this sort of problem will only get worse.

We should fix it and leave a FIXME to further weave mono error as we go.

**Pull request merged**